linkerd/metrics/src/summary.rs

```diff
@@ -0,0 +1,386 @@
+// This module is inspired by hdrhistogram-go, which is distributed under the
+// MIT license. Copyright (c) 2014 Coda Hale
+
+use crate::{Counter, Factor, FmtLabels, FmtMetric};
+pub use hdrhistogram::{AdditionError, CreationError, Histogram, RecordError};
+use parking_lot::{MappedRwLockWriteGuard, Mutex, MutexGuard, RwLock, RwLockWriteGuard};
+use std::fmt;
+use tokio::time;
+
+/// Summarizes a distribution of values at fixed quantiles over a sliding window.
+#[derive(Debug)]
+pub struct Summary<F = ()> {
+    rotate_interval: time::Duration,
+
+    windows: RwLock<Windows>,
```

Comment:
I'm not sure if this needs to be a `RwLock`, rather than just a `Mutex`. As far as I can tell, recording a value acquires a write lock on the `Windows`, while locking the `report` histogram acquires a read guard on the `Windows`. But, when the read guard is acquired, the `Mutex` around the report histogram is already locked. This means that the read lock is only acquired by one thread at a time, since the mutex around the report must be acquired before the read lock is acquired. Therefore, I don't think there are ever any scenarios where the read lock actually allows multiple readers to access this concurrently.

It might be better to just make this also be a mutex, and avoid the additional overhead of a rwlock?